### PR TITLE
EVG-6288: Spawn spot hosts with AWS Fleet API

### PR DESCRIPTION
**Pull request by ybrill**
The AWS Fleet API immediately gets the cheapest spot instance available. This way:
- we don't ask for the cheapest instances only to find out they aren't available and have to restart. 
- since the instances are returned immediately, this reduces code complexity needed to support the spot instance request lifecycle.

Some concerns:
- our EC2 Auto finds the cheaper of the lowest cost spot instance and the on-demand price. Since AWS Fleet will not give us _any_ hosts when the cost of the cheapest available spot hosts is >= on-demand hosts if the cheapest **reported** spot hosts are cheaper than on-demand but the cheapest **available** spot hosts are more expensive we will keep trying to get a spot host and failing.
- Deploying these changes will strand all the spot instance requests in progress at the time, since we will fail to get their status and update the host's id/status. Spot instances already running will not be affected.

### CODE REVIEW

**Comment by bsamek** [comment]
> * our EC2 Auto finds the cheaper of the lowest cost spot instance and the on-demand price. Since AWS Fleet will not give us _any_ hosts when the cost of the cheapest available spot hosts is >= on-demand hosts if the cheapest **reported** spot hosts are cheaper than on-demand but the cheapest **available** spot hosts are more expensive we will keep trying to get a spot host and failing.

This seems unlikely to happen, but is there any logging that we can add to report that this has happened?

> * Deploying these changes will strand all the spot instance requests in progress at the time, since we will fail to get their status and update the host's id/status. Spot instances already running will not be affected.

This could mean that we leak hosts. Is there something we can do, e.g., continuing to run the cloud host ready job for legacy hosts, in order to prevent this from happening?

File: cloud/ec2.go [hunk 1]
```diff
@@ -297,71 +278,112 @@ func (m *ec2Manager) spawnOnDemandHost(ctx context.Context, h *host.Host, ec2Set
...
+			{
+				LaunchTemplateSpecification: &ec2.FleetLaunchTemplateSpecificationRequest{
+					LaunchTemplateId: templateID,
+					Version:          aws.String(strconv.Itoa(int(*createTemplateResponse.LaunchTemplate.LatestVersionNumber))),
+				},
+			},
+		},
+		TargetCapacitySpecification: &ec2.TargetCapacitySpecificationRequest{
+			TotalTargetCapacity:       aws.Int64(1),
+			DefaultTargetCapacityType: aws.String("spot"),
```
Comment (bsamek) [inline cloud/ec2.go]: Is "spot" a constant in the `aws` package?

File: cloud/ec2.go [hunk 2]
```diff
@@ -297,71 +278,112 @@ func (m *ec2Manager) spawnOnDemandHost(ctx context.Context, h *host.Host, ec2Set
...
+		},
+		Type: aws.String("instant"),
```
Comment (bsamek) [inline cloud/ec2.go]: constant?

File: cloud/ec2.go [hunk 3]
```diff
@@ -297,71 +278,112 @@ func (m *ec2Manager) spawnOnDemandHost(ctx context.Context, h *host.Host, ec2Set
...
 
-	grip.Debug(message.Fields{
-		"message":       "starting spot instance",
-		"args":          spotRequest,
-		"host":          h.Id,
-		"host_provider": h.Distro.Provider,
-		"distro":        h.Distro.Id,
-	})
-	spotResp, err := m.client.RequestSpotInstances(ctx, spotRequest)
+	h.Id = *createFleetResponse.Instances[0].InstanceIds[0]
```
Comment (bsamek) [inline cloud/ec2.go]: Is it possible for the fleet response to _not_ have values at `Instances[0]` or `InstanceIds[0]`?

File: cloud/ec2.go [hunk 4]
```diff
@@ -776,90 +698,8 @@ func (m *ec2Manager) IsUp(ctx context.Context, h *host.Host) (bool, error) {
 	return false, nil
 }
 
-// OnUp is called when the host is up.
+// OnUp is a noop for ec2
 func (m *ec2Manager) OnUp(ctx context.Context, h *host.Host) error {
```
Comment (bsamek) [inline cloud/ec2.go]: Did you check that hosts start with the correct tags and can run tasks?

**Comment by ybrill** [comment]
@tychoish it is possible to have them exist side-by-side, as you're suggesting, and to define a new provider type for distros.

I don't understand the benefit, however: 
- The EC2-auto logic, as far as I can tell, is still necessary for Fleet, since Fleet does not return **any** hosts if Spot is more expensive than on-demand (instead of giving us an on-demand instance, as one would expect). The logic for an auto request still needs to be on our side.
- On-demand hosts don't benefit from being deployed through Fleet.

As I'm using Fleet, it's basically only a drop-in replacement for spot instance requests (once we've already decided we want a spot instance).

File: cloud/ec2.go [hunk 5]
```diff
@@ -297,71 +278,112 @@ func (m *ec2Manager) spawnOnDemandHost(ctx context.Context, h *host.Host, ec2Set
...
-			BlockDeviceMappings: blockDevices,
-		},
+func (m *ec2Manager) spawnSpotHost(ctx context.Context, h *host.Host, ec2Settings *EC2ProviderSettings, blockDevices []*ec2.LaunchTemplateBlockDeviceMappingRequest) ([]*string, error) {
+	// Cleanup
+	var templateID *string
+	defer func() {
+		if templateID != nil {
+			// Delete launch template
+			_, err := m.client.DeleteLaunchTemplate(ctx, &ec2.DeleteLaunchTemplateInput{LaunchTemplateId: templateID})
+			grip.ErrorWhen(err != nil, errors.Wrap(err, "can't delete launch template"))
```
[the remark this answers is not in the document]
Comment (ybrill) [reply cloud/ec2.go]: done.

File: cloud/ec2.go (hunk 1 above)
Comment (ybrill) [reply cloud/ec2.go]: yes. 
done.

File: cloud/ec2.go (hunk 2 above)
Comment (ybrill) [reply cloud/ec2.go]: yup :smile:

File: cloud/ec2.go (hunk 3 above)
Comment (ybrill) [reply cloud/ec2.go]: `validateCreateFleetResponse` makes sure this is not the case.

**Comment by tychoish** [comment]
I don't think it's reasonable/workable deploy this in a way that breaks deploys/reverts. The system needs to be compatible forwards/backwards, and we don't want to have to do extra manual work (or strand hosts) during the deploy, particularly if there's a revert. 

I'm also not sure if I totally buy the notion that we can't use this to start on-demand instances. The [documentation](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-fleet-configuration-strategies.html) contains the following line: 

> Determine what portion of the EC2 Fleet target capacity must be On-Demand capacity and Spot capacity. You can specify 0 for On-Demand capacity or Spot capacity, or both. 

If this isn't obvious in the documentation we should work with our contacts at AWS to figure out how to do this. 

(also I was reading through some of the documentation, and it looks like we can tag the instances during fleet creation, which might save a lot of AWS API traffic, and I think trying to reuse too much of the existing manager logic would make these kinds of improvements difficult to take advantage of.)

**Comment by ybrill** [comment]
Closing the PR for now to reimplement as a separate provider type.
Will wait until we hear back from AWS regarding EC2Auto-like functionality built into Fleet.

**Pull request closed without merging**